Rust: Flow through captured variables

paldepind opened the pull request.

paldepind commented:
The QLDoc checker is complaining, but it doesn't seem to be directly about the added code and otherwise the CI is happy.

intrigus-lgtm commented on rust/ql/lib/codeql/rust/dataflow/Ssa.qll:
````diff
@@ -338,4 +337,37 @@ module Ssa {
+   * An SSA definition inserted at a call that may update the value of a captured
+   * variable. For example, in
+   *
+   * ```rb
````
```suggestion
   * ```rust
```

intrigus-lgtm commented on rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:
```diff
@@ -681,6 +767,26 @@ final class SingletonContentSet extends ContentSet, TSingletonContentSet {
 
+class LambdaCallKind = Unit;
+
+/** Holds if `creation` is an expression that creates a lambda of kind `kind` for `c`. */
```
```suggestion
/** Holds if `creation` is an expression that creates a lambda of kind `kind`. */
```

paldepind commented:
Thanks @intrigus-lgtm. I've applied the suggestions :)

hvitved commented on rust/ql/lib/codeql/rust/dataflow/Ssa.qll:
````diff
@@ -210,23 +210,22 @@ module Ssa {
-     * end
-     * ```
+     * This is either the value in a direct assignment, `x = value`, or in a
+     * `let` statement, `let x = value`. Note that patterns on the rhs. are
````
rhs -> lhs?

hvitved commented on rust/ql/lib/codeql/rust/dataflow/Ssa.qll:
````diff
@@ -338,4 +337,37 @@ module Ssa {
+   * ```rust
+   * fn capture_mut() {
+   *   let mut y = 0;
+   *   (0..5).for_each(|| {
````
`||` -> `|x|`?

hvitved commented on rust/ql/lib/codeql/rust/dataflow/Ssa.qll:
```diff
@@ -338,4 +337,37 @@ module Ssa {
+   *
+   * a definition for `y` is inserted at the call to `for_each`.
+   */
+  class CapturedCallDefinition extends Definition, SsaImpl::UncertainWriteDefinition {
```
Presumably this class was only introduced to improve the `toString` representation? I suggest moving it into `SsaImpl` or making it private.

hvitved commented on rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:
```diff
@@ -101,11 +101,19 @@ final class ParameterPosition extends TParameterPosition {
   predicate isSelf() { this = TSelfParameterPosition() }
 
+  /**
+   * Holds if this position represents a reference to a lambda itself. Only
```
I think we should use the closure terminology in Rust instead of lambda.

hvitved commented on rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:
```diff
@@ -402,6 +445,37 @@ module Node {
     final override string toString() { result = PostUpdateNode.super.toString() }
   }
 
+  private class CapturePostUpdateNode extends PostUpdateNode, CaptureNode {
```
Nit: Move this down after the class `CaptureNode`.

paldepind replied:
What about moving `CaptureNode` up above `PostUpdateNode` instead? Having all the extensions of `PostUpdateNode` follow right after the abstract class makes it easier to see what it's domain is.

hvitved commented on rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:
```diff
@@ -735,6 +843,12 @@ module RustDataFlow implements InputSig<Location> {
+    or
+    node instanceof Node::ClosureParameterNode
+    or
+    node instanceof Node::ClosureArgumentNode
```
I don't think we should hide `ClosureArgumentNode`s

hvitved reviewed (changes requested):
Great work!

hvitved reviewed (changes requested):
One last nit.

paldepind commented:
I've now added the consistency checks for captured variables. I've fixed an inconsistency, which I think occurred because async blocks capture variables (due to how we've defined capture) but was not considered a lambda expression. So I've added a case for async blocks in `lambdaCreationExpr` (never mind what I said about closures with async blocks, that _might_ be a problem, but was not the cause of the inconsistency)

I've not added metric queries for the new consistency checks.

paldepind commented:
DCA shows new data flow inconsistencies. I'll investigate.

paldepind commented:
@hvitved Do you think it would be ok to merge the PR as-is and accept the additional data flow inconsistencies?

The inconsistencies doesn't seem trivial to fix and are of the form "Node steps to itself" and occur for some phi nodes at the entries to loops when there are both captured variables and `continue` statements.

Here's one example:
```rust
pub fn foo(cond: bool, names: Vec<Option<String>>) {
  let default_pattern = "hello";
  for name in names { // INCONSISTENCY: phi(default_pattern) | Node steps to itself
    if cond {
      let _ = name.unwrap_or_else(|| default_pattern.to_string());
      continue;
    }
  }
}
```

paldepind commented:
I've added a new test that reproduces the inconsistency and added an internal issue to track it.
